// grommet-controls/components/Spinning/doc/Spinning.stories.js
import React from 'react';
import { Box } from 'grommet';
import { Spinning } from '../Spinning';
export default {
    title: 'CONTROLS/controls/Spinning',
    component: Spinning,
};
export var main = function () { return React.createElement(Spinning, null); };
export var kind = function () { return (React.createElement(Box, { gap: 'medium' },
    React.createElement(Spinning, { kind: 'circle' }),
    React.createElement(Spinning, { kind: 'pulse' }),
    React.createElement(Spinning, { kind: 'three-bounce' }),
    React.createElement(Spinning, { kind: 'cube-grid' }),
    React.createElement(Spinning, { kind: 'wave' }),
    React.createElement(Spinning, { kind: 'folding-cube' }),
    React.createElement(Spinning, { kind: 'double-bounce' }),
    React.createElement(Spinning, { kind: 'wandering-cubes' }),
    React.createElement(Spinning, { kind: 'chasing-dots' }),
    React.createElement(Spinning, { kind: 'rotating-plane' }))); };
export var color = function () { return (React.createElement(Box, { gap: 'medium' },
    React.createElement(Spinning, { kind: 'circle', color: 'brand' }),
    React.createElement(Spinning, { kind: 'pulse', color: 'accent-1' }),
    React.createElement(Spinning, { kind: 'three-bounce', color: 'accent-2' }),
    React.createElement(Spinning, { kind: 'cube-grid', color: 'accent-3' }),
    React.createElement(Spinning, { kind: 'wave', color: 'neutral-1' }),
    React.createElement(Spinning, { kind: 'folding-cube', color: 'red' }),
    React.createElement(Spinning, { kind: 'double-bounce', color: 'blue' }),
    React.createElement(Spinning, { kind: 'wandering-cubes' }),
    React.createElement(Spinning, { kind: 'chasing-dots' }),
    React.createElement(Spinning, { kind: 'rotating-plane' }))); };
export var size = function () { return (React.createElement(Box, { gap: 'medium' },
    React.createElement(Spinning, { kind: 'circle', color: 'brand', size: 'xsmall' }),
    React.createElement(Spinning, { kind: 'pulse', color: 'accent-1', size: 'small' }),
    React.createElement(Spinning, { kind: 'circle', color: 'brand', size: 'medium' }),
    React.createElement(Spinning, { kind: 'three-bounce', color: 'accent-2', size: 'large' }),
    React.createElement(Spinning, { kind: 'cube-grid', color: 'accent-3', size: 'xlarge' }))); };
