// grommet-controls/components/Value/Value.js
import React from 'react';
import { Box, Text } from 'grommet';
var LABEL_SIZE_MAP = {
    xsmall: 'xsmall',
    small: 'xsmall',
    medium: 'xsmall',
    large: 'small',
    xlarge: 'medium',
    xxlarge: 'large',
};
/**
 * A single value visualization with a label<br/>
 * `import { Value } from 'grommet-controls';`<br/>
 * `<Value value={...} label={...} />`<br/>
 */
var Value = function (_a) {
    var label = _a.label, value = _a.value, size = _a.size, weight = _a.weight, gap = _a.gap, units = _a.units, color = _a.color;
    return (React.createElement(Box, { gap: gap },
        React.isValidElement(value) ? value : (React.createElement(Text, { size: size, weight: weight, color: color }, "" + (value !== undefined ? value : '-') + (units ? " " + units : ''))),
        React.createElement(Box, { alignSelf: 'end' },
            React.createElement(Text, { size: LABEL_SIZE_MAP[size], color: color }, label))));
};
Value.defaultProps = {
    label: undefined,
    value: undefined,
    weight: 'bold',
    gap: 'small',
    size: 'large',
    color: undefined,
};
export { Value };
