// grommet-controls/chartjs/BarChart/doc/BarChart.stories.js
import React from 'react';
import { BarChart } from '../BarChart';
import { rndDatasets } from '../../data/data';
export default {
    title: 'CHARTS/chartjs/BarChart',
    component: BarChart,
};
export var main = function () { return (React.createElement(BarChart, { data: {
        labels: ['January', 'February', 'March', 'April', 'May', 'June', 'July'],
        datasets: [
            {
                label: 'Dataset 1',
                data: [-40, 92, -44, -75, -65, -89, 78],
                borderWidth: 1,
            },
            {
                label: 'Dataset 2',
                data: [-78, -21, -43, 70, 11, -91, -63],
                borderWidth: 1,
            },
        ],
    } })); };
export var multiAxis = function () { return (React.createElement(BarChart, { data: rndDatasets(2, [{ yAxisID: 'y-axis-0' }, { yAxisID: 'y-axis-1' }]), options: {
        tooltips: {
            mode: 'index',
            intersect: true,
        },
        scales: {
            yAxes: [{
                    type: 'linear',
                    display: true,
                    position: 'left',
                    id: 'y-axis-0',
                }, {
                    type: 'linear',
                    display: true,
                    position: 'right',
                    id: 'y-axis-1',
                    gridLines: {
                        drawOnChartArea: false,
                    },
                }],
        },
    } })); };
export var stacked = function () { return (React.createElement(BarChart, { data: rndDatasets(3), options: {
        tooltips: {
            mode: 'index',
            intersect: true,
        },
        scales: {
            xAxes: [{
                    stacked: true,
                }],
            yAxes: [{
                    stacked: true,
                }],
        },
    } })); };
export var stackedGroup = function () { return (React.createElement(BarChart, { data: rndDatasets(3, [{ stack: 'stack-1' }, { stack: 'stack-2' }, { stack: 'stack-1' }]), options: {
        tooltips: {
            mode: 'index',
            intersect: true,
        },
        scales: {
            xAxes: [{
                    stacked: true,
                }],
            yAxes: [{
                    stacked: true,
                }],
        },
    } })); };
export var mixedLine = function () { return (React.createElement(BarChart, { data: rndDatasets(2, [{ type: 'line', fill: false }, { type: 'bar' }]) })); };
