// grommet-controls/components/Value/doc/Value.stories.js
import React from 'react';
import { Box, Text } from 'grommet';
import { Down } from 'grommet-icons';
import { Value } from '../Value';
export default {
    title: 'CONTROLS/controls/Value',
    component: Value,
};
export var main = function () { return (React.createElement(Box, { direction: 'row' },
    React.createElement(Box, { basis: 'small' },
        React.createElement(Value, { value: '30%', label: 'last quarter sales' })))); };
export var label = function () { return (React.createElement(Value, { value: '30%', label: 'a descriptive label' })); };
export var value = function () { return (React.createElement(Value, { value: (React.createElement(Box, { direction: 'row', align: 'center', gap: 'xsmall' },
        React.createElement(Text, { size: 'large', weight: 'bold', color: 'status-error' }, "30%"),
        React.createElement(Down, { color: 'status-error' }))), label: 'custom value' })); };
export var color = function () { return (React.createElement(Value, { label: 'status-ok', value: '30%', color: 'status-ok' })); };
export var weight = function () { return (React.createElement(Box, { direction: 'row' },
    React.createElement(Box, { basis: 'small', gap: 'small' },
        React.createElement(Value, { weight: 'normal', value: '30%', label: 'normal' }),
        React.createElement(Value, { weight: 'bold', value: '30%', label: 'bold' }),
        React.createElement(Value, { weight: 400, value: '30%', label: '400' })))); };
export var size = function () { return (React.createElement(Box, { direction: 'row' },
    React.createElement(Box, { basis: 'small', gap: 'small' },
        React.createElement(Value, { size: 'xsmall', value: '30%', label: 'xsmall' }),
        React.createElement(Value, { size: 'small', value: '30%', label: 'small' }),
        React.createElement(Value, { size: 'medium', value: '30%', label: 'medium' }),
        React.createElement(Value, { size: 'large', value: '30%', label: 'large' }),
        React.createElement(Value, { size: 'xlarge', value: '30%', label: 'xlarge' }),
        React.createElement(Value, { size: 'xxlarge', value: '30%', label: 'xxlarge' })))); };
