// grommet-controls/components/DateInputField/doc/DateInputField.stories.js
import React from 'react';
import { Box, Button } from 'grommet';
import { action } from '@storybook/addon-actions';
import { DateInputField } from '../DateInputField';
import { Form, validators, } from '../../../index';
import { smallDate } from '../../../utils/moment';
export default {
    title: 'CONTROLS/validation/DateInputField',
    component: DateInputField,
};
export var main = function () { return (React.createElement(Form, { onSubmit: action('onSubmit'), pad: { horizontal: 'small' }, focusFirstChild: false },
    React.createElement(DateInputField, { name: 'date', label: 'Date', validation: [validators.required()] }),
    React.createElement(Box, { pad: 'small' },
        React.createElement(Button, { type: 'submit', label: 'Submit' })))); };
export var inField = function () { return (React.createElement(Form, { onSubmit: action('onSubmit'), pad: { horizontal: 'small' }, focusFirstChild: false },
    React.createElement(DateInputField, { name: 'date', label: 'Date', inField: false, validation: [validators.required()] }),
    React.createElement(Box, { pad: 'small' },
        React.createElement(Button, { type: 'submit', label: 'Submit' })))); };
export var maskFormat = function () {
    var form = React.useState({ date: smallDate(new Date(), 'es-ES') })[0];
    return (React.createElement(Form, { onSubmit: action('onSubmit'), pad: { horizontal: 'small' }, focusFirstChild: false, object: form },
        React.createElement(DateInputField, { name: 'date', autocorrect: true, maskFormat: 'dd mm yyyy', locale: 'es-ES', label: 'Ingreso', validation: [validators.required('Seleccione fecha de ingreso')] }),
        React.createElement(Box, { pad: 'small' },
            React.createElement(Button, { type: 'submit', label: 'Submit' }))));
};
export var validation = function () { return (React.createElement(Form, { onSubmit: action('onSubmit'), pad: { horizontal: 'small' }, focusFirstChild: false },
    React.createElement(DateInputField, { name: 'date', label: 'Date', validation: [
            validators.required(),
            function (_row, value) {
                if (value) {
                    var date = new Date(value);
                    if (date >= new Date('01/01/2000')) {
                        return 'date must be before 01/01/2000';
                    }
                }
                return undefined;
            },
        ] }),
    React.createElement(Box, { pad: 'small' },
        React.createElement(Button, { type: 'submit', label: 'Submit' })))); };
