// grommet-controls/components/Colors/doc/Colors.stories.js
import React from 'react';
import { Box } from 'grommet';
import { action } from '@storybook/addon-actions';
import { Colors } from '../Colors';
import { materialColors, uiColors, basicColors, } from '../../../index';
export default {
    title: 'CONTROLS/controls/Colors',
    component: Colors,
};
var blue = {
    '50': '#e3f2fd', '100': '#bbdefb', '200': '#90caf9', '300': '#64b5f6', '400': '#42a5f5', '500': '#2196f3', '600': '#1e88e5', '700': '#1976d2', '800': '#1565c0', '900': '#0d47a1', 'a100': '#82b1ff', 'a200': '#448aff', 'a400': '#2979ff', 'a700': '#2962ff',
};
var cyan = {
    '50': '#e0f7fa',
    '100': '#b2ebf2',
    '200': '#80deea',
    '300': '#4dd0e1',
    '400': '#26c6da',
    '500': '#00bcd4',
    '600': '#00acc1',
    '700': '#0097a7',
    '800': '#00838f',
    '900': '#006064',
    'a100': '#84ffff',
    'a200': '#18ffff',
    'a400': '#00e5ff',
    'a700': '#00b8d4',
};
var deepOrange = {
    '50': '#fbe9e7',
    '100': '#ffccbc',
    '200': '#ffab91',
    '300': '#ff8a65',
    '400': '#ff7043',
    '500': '#ff5722',
    '600': '#f4511e',
    '700': '#e64a19',
    '800': '#d84315',
    '900': '#bf360c',
    'a100': '#ff9e80',
    'a200': '#ff6e40',
    'a400': '#ff3d00',
    'a700': '#dd2c00',
};
var brown = {
    '50': '#efebe9',
    '100': '#d7ccc8',
    '200': '#bcaaa4',
    '300': '#a1887f',
    '400': '#8d6e63',
    '500': '#795548',
    '600': '#6d4c41',
    '700': '#5d4037',
    '800': '#4e342e',
    '900': '#3e2723',
};
var blueGrey = {
    '50': '#eceff1',
    '100': '#cfd8dc',
    '200': '#b0bec5',
    '300': '#90a4ae',
    '400': '#78909c',
    '500': '#607d8b',
    '600': '#546e7a',
    '700': '#455a64',
    '800': '#37474f',
    '900': '#263238',
};
export var main = function () { return (React.createElement(Box, { direction: 'row' },
    React.createElement(Box, { basis: 'medium' },
        React.createElement(Colors, { size: 'small', onSelect: action('onSelect'), colors: materialColors })))); };
export var colors = function () { return (React.createElement(Colors, { size: 'small', color: '#ff0000', colors: uiColors })); };
export var onSelect = function () { return (React.createElement(Colors, { size: 'small', colors: basicColors, onSelect: action('onSelect') })); };
export var size = function () { return (React.createElement(Box, { gap: 'small' },
    React.createElement(Colors, { size: 'small', colors: {
            blue: blue, blueGrey: blueGrey, brown: brown, cyan: cyan, deepOrange: deepOrange,
        } }),
    React.createElement(Colors, { size: 'medium', columns: 5, colors: {
            blue: blue, blueGrey: blueGrey, brown: brown, cyan: cyan, deepOrange: deepOrange,
        } }),
    React.createElement(Colors, { size: 'large', columns: 3, colors: {
            blue: blue, blueGrey: blueGrey, brown: brown, cyan: cyan, deepOrange: deepOrange,
        } }))); };
export var columns = function () { return (React.createElement(Colors, { size: 'small', columns: 5, colors: {
        blue: blue, blueGrey: blueGrey, brown: brown, cyan: cyan, deepOrange: deepOrange,
    } })); };
export var wrap = function () { return (React.createElement(Colors, { size: 'small', columns: 5, wrap: true, colors: {
        blue: blue, blueGrey: blueGrey, brown: brown, cyan: cyan, deepOrange: deepOrange,
    } })); };
