// grommet-controls/components/MaskedInput/doc/MaskedInput.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { Box, Calendar } from 'grommet';
import { Add, Subtract } from 'grommet-icons';
import { MaskedInput } from '../MaskedInput';
import { smallDate } from '../../../utils/moment';
export default {
    title: 'CONTROLS/input/MaskedInput',
    component: MaskedInput,
};
var MaskedTestBed = function (_a) {
    var _b = _a.value, defaultValue = _b === void 0 ? '' : _b, props = __rest(_a, ["value"]);
    var _c = React.useState(defaultValue), value = _c[0], setValue = _c[1];
    return (React.createElement(Box, { direction: 'row' },
        React.createElement(Box, { basis: 'medium', gap: 'small' },
            React.createElement(MaskedInput, __assign({ value: value, onChange: function (_a) {
                    var target = _a.target;
                    return setValue(target.value);
                } }, props)))));
};
var DateInput = function (props) {
    var _a = React.useState(smallDate(new Date())), date = _a[0], setDate = _a[1];
    return (React.createElement(Box, { direction: 'row' },
        React.createElement(Box, { basis: 'medium' },
            React.createElement(MaskedInput, __assign({ mask: [/\d/, /\d/, '/', /\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/], dropContent: (React.createElement(Box, { pad: 'small' },
                    React.createElement(Calendar, { size: 'small', date: date, onSelect: function (isoDate) { return setDate(smallDate(new Date(isoDate))); } }))), value: date, onChange: function (_a) {
                    var target = _a.target;
                    return setDate(target.value);
                } }, props)))));
};
var NumberInput = function () {
    var _a = React.useState(10), number = _a[0], setNumber = _a[1];
    return (React.createElement(Box, { direction: 'row' },
        React.createElement(Box, { basis: 'medium' },
            React.createElement(MaskedInput, { value: number, onChange: function (_a) {
                    var value = _a.target.value;
                    return setNumber(parseFloat(value));
                }, mask: MaskedInput.createNumberMask({
                    allowDecimal: true,
                }), widgets: [
                    {
                        icon: React.createElement(Add, null),
                        onClick: function () { return setNumber(number + 1); },
                    },
                    {
                        icon: React.createElement(Subtract, null),
                        onClick: function () { return setNumber(number - 1); },
                    },
                ] }))));
};
export var main = function () { return (React.createElement(MaskedTestBed, { value: '8024442131', mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/] })); };
export var a11yTitle = function () { return (React.createElement(MaskedTestBed, { a11yTitle: 'Dollars', mask: MaskedInput.createNumberMask({}), value: '18933.85' })); };
export var disabled = function () { return (React.createElement(DateInput, { disabled: true })); };
export var dropContent = function () { return React.createElement(DateInput, null); };
export var widgets = function () { return React.createElement(NumberInput, null); };
export var plain = function () { return (React.createElement(MaskedTestBed, { value: '8024442131', mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/], plain: true })); };
export var focusIndicator = function () { return (React.createElement(MaskedTestBed, { value: '8024442131', mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/], plain: true, focusIndicator: true })); };
export var mask = function () { return (React.createElement(MaskedTestBed, { placeholder: 'US phone number with country code', mask: ['+', '1', ' ', '(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/] })); };
export var guide = function () { return (React.createElement(MaskedTestBed, { value: '8024442131', guide: false, mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/], plain: true, focusIndicator: true })); };
export var pipe = function () { return (React.createElement(MaskedTestBed, { mask: [MaskedInput.alphabetic, MaskedInput.digit, MaskedInput.alphabetic, ' ', MaskedInput.digit, MaskedInput.alphabetic, MaskedInput.digit], pipe: function (conformedValue) { return ({
        value: conformedValue.toUpperCase(),
    }); }, placeholder: 'K1A 0B2', placeholderChar: MaskedInput.placeholderChars.underscore })); };
export var placeholderChar = function () { return (React.createElement(MaskedTestBed, { mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/], placeholderChar: MaskedInput.placeholderChars.underscore, value: '8024442131' })); };
export var keepCharPositions = function () { return (React.createElement(MaskedTestBed, { mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/], keepCharPositions: true, value: '8024442131' })); };
export var showMask = function () { return (React.createElement(MaskedTestBed, { mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/], showMask: true, value: '8024442131' })); };
export var placeholder = function () { return (React.createElement(MaskedTestBed, { mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/], placeholder: 'Enter phone...', value: '8024442131' })); };
export var name = function () { return (React.createElement(MaskedTestBed, { mask: ['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/], id: 'date-id', name: 'date-name', value: '8024442131' })); };
