// grommet-controls/components/Tag/doc/Tag.stories.js
import React from 'react';
import { Github, Grommet, FormSubtract } from 'grommet-icons';
import { action } from '@storybook/addon-actions';
import { Tag } from '../Tag';
export default {
    title: 'CONTROLS/controls/Tag',
    component: Tag,
};
var tagIcons = [React.createElement(Github, null), React.createElement(Grommet, null)];
export var main = function () { return (React.createElement(Tag, { label: 'Tag', onChange: action('onChange') })); };
export var a11yTitle = function () { return (React.createElement(Tag, { a11yTitle: 'Tag component' })); };
export var icon = function () { return (React.createElement(Tag, { icon: React.createElement(FormSubtract, null) })); };
export var disabled = function () { return (React.createElement(Tag, { disabled: true, label: 'Disabled' })); };
export var focusable = function () { return (React.createElement(Tag, { focusable: false, label: 'Cant focus' })); };
export var background = function () { return (React.createElement(Tag, { background: 'accent-1', label: 'accent-1' })); };
export var border = function () { return (React.createElement(Tag, { border: {
        side: 'all', size: 'medium', color: 'accent-1',
    }, round: 'medium', label: 'border' })); };
export var round = function () { return (React.createElement(Tag, { round: 'medium', background: 'accent-2', label: 'round' })); };
export var onClick = function () { return (React.createElement(Tag, { onClick: action('onClick') })); };
export var onChange = function () {
    var _a = React.useState(0), iconIndex = _a[0], setIconIndex = _a[1];
    var onToggle = function () {
        setIconIndex(1 - iconIndex);
    };
    return (React.createElement(Tag, { icon: tagIcons[iconIndex], onClick: action('onClick'), onChange: onToggle }));
};
export var reverse = function () { return (React.createElement(Tag, { reverse: true, label: 'Adam' })); };
export var size = function () { return (React.createElement(Tag, { size: 'large', label: 'Large' })); };
