// grommet-controls/components/Card/doc/Card.stories.js
import React from 'react';
import { Box, Button, Heading, Image, Text, } from 'grommet';
import { Add, Bitcoin, Edit } from 'grommet-icons';
import { Card } from '../Card';
import { ImageStamp } from '../../ImageStamp';
export default {
    title: 'CONTROLS/controls/Card',
    component: Card,
};
var actions = [
    {
        label: 'Edit',
        icon: React.createElement(Edit, null),
        onClick: function () { },
    }, {
        label: 'New',
        icon: React.createElement(Add, null),
        onClick: function () { },
    },
];
export var main = function () { return (React.createElement(Card, null,
    React.createElement(Card.CardTitle, { border: 'bottom', align: 'center', direction: 'column', pad: 'small' },
        React.createElement(Box, { direction: 'row' },
            React.createElement(ImageStamp, { round: 'full', src: '//v2.grommet.io/assets/Wilderpeople_Ricky.jpg' }),
            React.createElement(Heading, { level: 3, margin: 'xsmall' }, "Info card")),
        React.createElement(Box, { direction: 'row' },
            React.createElement(Bitcoin, { color: 'plain' }),
            React.createElement(Text, null, "sub-title info"))),
    React.createElement(Card.CardContent, null,
        React.createElement(Image, { fit: 'cover', src: '//v2.grommet.io/assets/Wilderpeople_Ricky.jpg', height: '250' })),
    React.createElement(Card.CardActions, { justify: 'center' }, actions.map(function (action, index) { return (React.createElement(Button, { key: "actions_" + index, label: action.label, onClick: action.onClick })); })))); };
export var animation = function () { return (React.createElement(Card, { animation: {
        type: 'zoomIn', duration: 5000, size: 'xlarge',
    } },
    React.createElement(Card.CardTitle, { border: 'bottom' }, "Card"),
    React.createElement(Card.CardContent, null,
        React.createElement(Image, { fit: 'cover', src: '//v2.grommet.io/assets/Wilderpeople_Ricky.jpg', height: '250' })))); };
export var background = function () { return (React.createElement(Card, { background: 'accent-1' },
    React.createElement(Card.CardTitle, { border: 'bottom' }, "Card"),
    React.createElement(Card.CardContent, { basis: 'small' }, "Card content"))); };
export var border = function () { return (React.createElement(Card, { border: {
        color: 'brand', size: 'medium',
    } },
    React.createElement(Card.CardTitle, { border: 'bottom' }, "Card"),
    React.createElement(Card.CardContent, { basis: 'small' }, "Card content"))); };
export var elevation = function () { return (React.createElement(Card, { elevation: 'large' },
    React.createElement(Card.CardTitle, { border: 'bottom' }, "Card"),
    React.createElement(Card.CardContent, { basis: 'small' }, "Card content"))); };
export var gap = function () { return (React.createElement(Card, { gap: 'medium' },
    React.createElement(Card.CardTitle, { border: 'bottom' }, "Card"),
    React.createElement(Card.CardContent, { basis: 'small' }, "Card content"))); };
export var pad = function () { return (React.createElement(Card, { pad: 'medium' },
    React.createElement(Card.CardTitle, { border: 'bottom' }, "Card"),
    React.createElement(Card.CardContent, { basis: 'small' }, "Card content"))); };
export var round = function () { return (React.createElement(Card, { round: 'medium', pad: 'small' },
    React.createElement(Card.CardTitle, { border: 'bottom' }, "Card"),
    React.createElement(Card.CardContent, { basis: 'small' }, "Card content"))); };
